pwnlib/encoders/i386/ascii_shellcode.py

```diff
@@ -0,0 +1,300 @@
+         subtracted from `last` will equal `target` while only constructing
+         integers from bytes in `vocab`
+
+        int_size is take from the context (context.bits / 8)
+
+        Args:
+            last (bytes): Current value of eax
+            target (bytes): Desired value of eax
+            vocab (bytes): Allowed characters
+            max_subs (int): Maximum subtraction attempts
+
+        Raises:
+            ArithmeticError: If a sequence of subtractions could not be found
+
+        Returns:
+            List[bytearray]: List of numbers that would need to be subtracted
+            from `last` to get to `target`
+
+        Examples:
+
+            >>> context.update(arch='i386', os='linux')
+            >>> vocab = b'!"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
+            >>> print(encoders.i386.ascii_shellcode.encode._calc_subtractions(b'\x10'*4, b'\x11'*4, vocab))
+            [bytearray(b'!!!!'), bytearray(b'`___'), bytearray(b'~~~~')]
+            >>> print(encoders.i386.ascii_shellcode.encode._calc_subtractions(b'\x11\x12\x13\x14', b'\x15\x16\x17\x18', vocab))
+            [bytearray(b'~}}}'), bytearray(b'~~~~')]
+        """
+        int_size = context.bits // 8
+        subtractions = [bytearray(b'\x00'*int_size)]
+        if six.PY2:
+            last = map(ord, last)
+            target = map(ord, target)
+        for subtraction in range(max_subs):
+            carry = success_count = 0
+            for byte in range(int_size):
+                # Try all combinations of all the characters in vocab of
+                # `subtraction` characters in each combination. So if `max_subs`
+                # is 4 and we're on the second subtraction attempt, products will
+                # equal [\, ", #, %, ...], [\, ", #, %, ...], 0, 0
+                for products in product(
+                    *[x <= subtraction and vocab or (0,) for x in range(max_subs)]
+                ):
+                    if six.PY2:
+                        products = map(lambda x: isinstance(x, str) and ord(x) or x, products)
+                    # Sum up all the products, carry from last byte and the target
+                    attempt = sum(chain(
+                        (target[byte], carry), products
+                    ))
+                    # If the attempt equals last, we've found the combination
+                    if last[byte] == attempt & 0xff:
+                        carry = (attempt & 0xff00) >> 8
+                        # Update the result with the current `products`
+                        for p, i in zip(products, range(subtraction+1)):
+                            subtractions[i][byte] = p
+                        success_count += 1
+                        break
+            if success_count == int_size:
+                return subtractions
+            else:
+                subtractions.append(bytearray(b'\x00'*int_size))

```

Comment:
This would be enough:
```suggestion
                subtractions.append(bytearray(int_size))
```